### PR TITLE
gh-111965: Use critical sections to make io.TextIOWrapper thread safe

### PR DESCRIPTION
**Pull request by aisk**
* Issue: gh-111965

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2023-11-17-17-34-30.gh-issue-111965.6iMWn0.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Using critical section to make ``io.TextIOWrapper`` thread safe without GIL.
```
Comment (corona10) [inline Misc/NEWS.d/next/Library/2023-11-17-17-34-30.gh-issue-111965.6iMWn0.rst]: As I commented before, NEWS.d is not needed for this task

Comment (aisk) [reply Misc/NEWS.d/next/Library/2023-11-17-17-34-30.gh-issue-111965.6iMWn0.rst]: Sorry, just noticed it

**Review by colesbury — APPROVED** [review]
Thanks!

@corona10 would you please look over this one as well?

File: Modules/_io/textio.c [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #include "Python.h"
 #include "pycore_call.h"          // _PyObject_CallMethod()
 #include "pycore_codecs.h"        // _PyCodecInfo_GetIncrementalDecoder()
+#include "pycore_critical_section.h"
```
Comment (corona10) [inline Modules/_io/textio.c]: Please add a comment as the same as the other header. `// Py_BEGIN_CRITICAL_SECTION()`

**Review by corona10 — CHANGES_REQUESTED** [review]

File: Modules/_io/textio.c (hunk 2 above)
Comment (aisk) [reply Modules/_io/textio.c]: Updated

File: Modules/_io/textio.c [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #include "Python.h"
 #include "pycore_call.h"          // _PyObject_CallMethod()
 #include "pycore_codecs.h"        // _PyCodecInfo_GetIncrementalDecoder()
+#include "pycore_critical_section.h"  // Py_BEGIN_CRITICAL_SECTION()
 #include "pycore_fileutils.h"     // _Py_GetLocaleEncoding()
 #include "pycore_interp.h"        // PyInterpreterState.fs_codec
 #include "pycore_long.h"          // _PyLong_GetZero()
```
Comment (corona10) [inline Modules/_io/textio.c]: nit: Please lint comments in following way :)
(Please check on your local environment too)
```suggestion
#include "pycore_call.h"              // _PyObject_CallMethod()
#include "pycore_codecs.h"            // _PyCodecInfo_GetIncrementalDecoder()
#include "pycore_critical_section.h"  // Py_BEGIN_CRITICAL_SECTION()
#include "pycore_fileutils.h"         // _Py_GetLocaleEncoding()
#include "pycore_interp.h"            // PyInterpreterState.fs_codec
#include "pycore_long.h"              // _PyLong_GetZero()
```

**Review by corona10 — APPROVED** [review]
lgtm

**Comment by aisk** [comment]
Thanks for the review!

**Pull request merged**